<a href="https://colab.research.google.com/github/j20nch/class2022Spring/blob/main/Sentence_similarity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install gradio
import gradio as gr

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 5.1 MB 7.4 MB/s 
     |████████████████████████████████| 256 kB 68.4 MB/s 
     |████████████████████████████████| 84 kB 2.8 MB/s 
     |████████████████████████████████| 212 kB 26.2 MB/s 
     |████████████████████████████████| 2.0 MB 29.1 MB/s 
     |████████████████████████████████| 1.1 MB 40.3 MB/s 
     |████████████████████████████████| 54 kB 2.5 MB/s 
     |████████████████████████████████| 53 kB 1.9 MB/s 
     |████████████████████████████████| 271 kB 59.7 MB/s 
     |████████████████████████████████| 144 kB 48.4 MB/s 
     |████████████████████████████████| 94 kB 3.6 MB/s 
     |████████████████████████████████| 11.1 MB 45.2 MB/s 
     |████████████████████████████████| 63 kB 1.8 MB/s 
     |████████████████████████████████| 80 kB 8.6 MB/s 
     |████████████████████████████████| 43 kB 1.9 MB/s 
     |████████████████████████████████| 4.0 M

Sentence Similarity: 주어진 두 문장들이 얼마나 유사한지를 보는 것. / Two texts. paragraph든, 문장이 두 개든 상관 없음. Two text가 얼마나 유사한지를 determine해주는 것이다! 처음에 모든 모델들은 input 자체(text 등)를 바로 받아들일 수가 없음. 어떤 식으로든 "중간 단계"가 필요함. feature라고도 하는데, vector라고도 하고, embeddings라고도 함. text, sound, image 등의 raw data 자체를 모델이 받아들이기는 힘듦. 그래서 feature의 형태로 한 번 convert를 해 줘야 함! 

왜 vector, embeddings로 바꿔주나? 그래야만 두 텍스트 간 similarity를 쉽게 계산할 수 있기 때문.

In [2]:
!pip install sentence_transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 79 kB 4.4 MB/s 
     |████████████████████████████████| 4.2 MB 16.7 MB/s 
     |████████████████████████████████| 1.2 MB 41.2 MB/s 
     |████████████████████████████████| 86 kB 4.7 MB/s 
     |████████████████████████████████| 6.6 MB 36.4 MB/s 
     |████████████████████████████████| 596 kB 39.0 MB/s 
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.0-py3-none-any.whl size=120747 sha256=3378e3107720ad989dfa47d3a9bce5b3de4b0466d560042cf87291b031956b42
  Stored in directory: /root/.cache/pip/wheels/83/c0/df/b6873ab7aac3f2465aa9144b6b4c41c4391cfecc027c8b07e7
Successfully built sentence-transformers
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [3]:
from sentence_transformers import SentenceTransformer, util
sentences = ["This is an example sentence", "it is one example writing"]

model = SentenceTransformer('sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2')
embeddings = model.encode(sentences)
print(embeddings)

Downloading:   0%|          | 0.00/968 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/190 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/3.79k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/645 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/122 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/229 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/471M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/5.07M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/239 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/9.08M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/480 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/14.8M [00:00<?, ?B/s]

[[ 1.88456848e-01  1.74256176e-01  5.44777215e-02  2.90517420e-01
   1.67664081e-01 -4.72069457e-02  6.45580173e-01  1.59808800e-01
   2.26892605e-01 -3.08905598e-02  2.55883545e-01 -5.25876768e-02
  -2.26101175e-01 -5.71063384e-02  1.30426243e-01  1.24953501e-01
   3.17496210e-01  1.94444105e-01 -5.86325705e-01 -1.25861000e-02
   6.09909177e-01  1.64327413e-01  3.33113708e-02 -2.73830682e-01
  -2.89757580e-01 -2.11196810e-01 -2.26138867e-02 -1.70359373e-01
   1.61589950e-01  6.08273745e-02 -2.41624042e-01  1.85792103e-01
   4.27409589e-01  1.92951798e-01 -7.23447800e-02  1.66111022e-01
   1.04428008e-01  2.04772323e-01  2.11167201e-01  1.99740008e-01
  -9.40827355e-02 -1.73836619e-01  6.42733201e-02  2.80254930e-01
  -2.95305759e-01  6.20952919e-02  1.04276791e-01 -2.36439407e-02
   1.29131526e-01 -1.26174763e-01 -1.78990051e-01  3.70057598e-02
  -6.12506211e-01  5.02983071e-02  1.77303478e-01  2.24941239e-01
   1.73860714e-01 -3.84028591e-02 -2.12868378e-01  2.58492470e-01
  -1.21016

두 비슷한 문장이 주어져 있고, 그게 얼마만큼 비슷한지를 계산해줌. 
sentences: 두 문장이 들어간 list
model: 모델. 
embeddings: 위 모델을 가지고 입력으로 위의 문장들을 쓰면 embeddings를 뱉어준다! print out을 해보니 숫자들이 무수히 많이 나옴. 
> 이게 무슨 뜻일까?
: embeddings의 구조를 보면 list가 두 개 들어가 있음. 첫번째 list는 첫번째 text, 두번째 list는 두번째 text. 그걸 list로 감싸고 있음.


In [4]:
len(embeddings[1])

384

len(embeddings)=2, len(embeddings[0])=len(embeddings[1]=384) (embeddings for text1, embeddings for text 2)
> 384개의 숫자가 한 list안에 들어있다! 

이것들을 feature, feature vector라고도 부름. vector는 data, 숫자들이라 했는데 이것도 마찬가지로 숫자 data. 

그럼 이 2개를 가지고 뭘 하냐? Similarity를 구할 수 있음.

<아주 중요한 부분>
ex) 2개의 vector가 있음. 1: [1, 2] / 2: [2, 1]
왼쪽 걸 x축, 오른쪽 걸 y축이라 하면 그걸 찍어볼 수 있음. (좌표 상에서 (1,2)와 (2,1))

> 이 2개가 얼마만큼 유사한지는 (0,0), (1,2)의 선과 (0,0), (2,1)의 선 사이의 각도 theta를 통해 알 수 있음. 이 각도가 0도라면 유사도가 제일 높고, 각도가 90도라면 유사도가 제일 낮다!
>> 거기에 cosine을 붙임. cos(0)=1, cos(90)=0(=cos(파이/2)), 그래서 similarity between two vectors(embeddings)는 두 vector의 각도 값의 cosine이다! 이걸 "Cosine Similarity"라 이야기 함. (cosine을 적용시키면 -1에서 1사이의 숫자로 도출될 것)


* 지금 숫자가 2개라 이 각도가 됨. 3개면? 3차원의 각도, 4개면 4차원의 각도. 각도는 1개밖에 없음. 숫자가 많아져서 100차원이 된다고 한들, 원점하고 첫번째 vector, 두번째 vector 간에는 평면삼각형이 나오게 되어 있음! 위 예시도 384차원인데, 두 vector는 어짜피 삼각형을 만들게 되어 있음!

In [5]:
cosine_scores = util.pytorch_cos_sim(embeddings[0], embeddings[1])
cosine_scores

tensor([[0.5726]])

위에서 설명한 것을 계산해주는 것. 두 vector(첫번째, 두번째 문장에 대한 각각 384개의 숫자)가 이루는 각도의 cosine값을 계산.

In [7]:
def func (text1, text2):
  from sentence_transformers import SentenceTransformer, util
  model = SentenceTransformer('sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2')
  embeddings = model.encode([text1, text2])
  cosine_scores = util.pytorch_cos_sim(embeddings[0], embeddings[1])
  return cosine_scores

In [8]:
examples = [["This is an example sentence", "it is one example writing"], ["A frog is hopping near the pond", "I love Korean Food"]]

In [9]:
gr.Interface(fn=func, inputs = ['text','text'], outputs = 'number', examples = examples).launch()

Colab notebook detected. To show errors in colab notebook, set `debug=True` in `launch()`
Running on public URL: https://44501.gradio.app

This share link expires in 72 hours. For free permanent hosting, check out Spaces (https://huggingface.co/spaces)


(<gradio.routes.App at 0x7f7c986db250>,
 'http://127.0.0.1:7860/',
 'https://44501.gradio.app')

-1에서 1까지 간다고 쳤을 때, 0에 가까울 수록 2개의 similarity는 작음!